vermouth/ffinput.py

```diff
@@ -51,196 +55,337 @@
+    @SectionLineParser.section_parser('link', '!bonds', context_type='link')
+    @SectionLineParser.section_parser('link', '!angles', context_type='link')
+    @SectionLineParser.section_parser('link', '!dihedrals', context_type='link')
+    @SectionLineParser.section_parser('link', '!constraints', context_type='link')
+    @SectionLineParser.section_parser('link', '!pairs', context_type='link')
+    @SectionLineParser.section_parser('link', '!exclusions', context_type='link')
+    @SectionLineParser.section_parser('link', '!virtual_sites2', context_type='link')
+    @SectionLineParser.section_parser('link', '!virtual_sitesn', context_type='link')
+    @SectionLineParser.section_parser('modification', 'bonds', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'angles', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'dihedrals', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'constraints', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'pairs', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'exclusions', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'virtual_sites2', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'impropers', context_type='modification')
+    @SectionLineParser.section_parser('modification', 'virtual_sitesn', context_type='modification')
+    def _interactions(self, line, lineno=0, context_type=''):
+        context = self.get_context(context_type)
+        interaction_name = self.section[-1]
+        delete = False
+        if interaction_name.startswith('!'):
+            interaction_name = interaction_name[1:]
+            delete = True
+        tokens = collections.deque(_tokenize(line))
+        if tokens[0] == '#meta':
+            _parse_meta(
+                tokens,
+                context,
+                context_type=context_type,
+                section=interaction_name,
+            )
+        else:
+            n_atoms = self.interactions_natoms.get(interaction_name)
+            _base_parser(
+                tokens,
+                context,
+                context_type=context_type,
+                section=interaction_name,
+                natoms=n_atoms,
+                delete=delete,
+            )
+
+    @SectionLineParser.section_parser('moleculetype', 'patterns')
+    @SectionLineParser.section_parser('moleculetype', 'features')
+    @SectionLineParser.section_parser('moleculetype', 'non-edge')
+    @SectionLineParser.section_parser('modifications', 'non-edge')
+    def _invalid_out_of_link(self, line, lineno=0):
+        raise IOError('The "{}" section is only valid in links.'
+                      .format(self.section[-1]))
+
+    @SectionLineParser.section_parser('link', context_type='link')
```

Comment:
call _new_link here. Same comment as with _new_block above

Reply:
This will not work, because there are instances where you have:
`[link]`
`[atoms]`

In that case the method is not called.